### PR TITLE
Widget mode apps with data now work

### PR DESCRIPTION
**Pull request by jmkulwik**
# Description
In the past, widget mode that used data wouldn't work on the first page load. This was because the roundtrip to firebase for level data wouldn't finish before the run button was auto-clicked to start the widget mode level. This change updates applab to wait for the roundtrip to firebase to finish before running the widget mode level.

Sample level: https://studio.code.org/s/csd1-pilot/stage/3/puzzle/2

image

### Future work
As part of this, it became apparent that our Applab init function needs refactoring with specific regard toward timing issues like this one. I've added a future work item to handle this tech debt.
https://codedotorg.atlassian.net/browse/STAR-721

## Links

- [jira](https://codedotorg.atlassian.net/browse/STAR-660)

## Testing story
Updated old tests to use the new version of PopulateTables that uses a promise rather than callbacks.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/storage/firebaseStorage.js [hunk 1]
```diff
@@ -628,18 +628,13 @@ function getRecordsData(records) {
  * @param {function ()} onSuccess Function to call on success.
  * @param {function} onError Function to call with an error in case of failure.
```
Comment (ajpal) [inline apps/src/storage/firebaseStorage.js]: Please update jsdoc

File: apps/test/unit/storage/FirebaseStorageTest.js [hunk 2]
```diff
@@ -895,9 +897,11 @@ describe('FirebaseStorage', () => {
             .set(EXISTING_COUNTER_DATA)
         )
         .then(() => {
-          FirebaseStorage.populateTable(
+          let firebasePromise = FirebaseStorage.populateTable(
```
Comment (ajpal) [inline apps/test/unit/storage/FirebaseStorageTest.js]: Does this need to be a variable? Could you just `FirebaseStorage.populateTable(...).then(...)` (and below)

File: apps/src/applab/applab.js [hunk 3]
```diff
@@ -772,9 +770,14 @@ Applab.init = function(config) {
           );
         }
       }
-
+    })
+    .then(() => {
       if (getStore().getState().pageConstants.widgetMode) {
-        Applab.runButtonClick();
+        dataLoadedPromise
```
Comment (ajpal) [inline apps/src/applab/applab.js]: can you set `dataLoadedPromise = Promise.resolve()` if there are no tables to load? That way you don't need this ternary operator here, you can just do `dataLoadedPromise.then(...)`

File: apps/src/storage/firebaseStorage.js (hunk 1 above)
Comment (jmkulwik) [reply apps/src/storage/firebaseStorage.js]: Thanks for catching that!

File: apps/test/unit/storage/FirebaseStorageTest.js (hunk 2 above)
Comment (jmkulwik) [reply apps/test/unit/storage/FirebaseStorageTest.js]: Good point. Thanks!

File: apps/src/applab/applab.js (hunk 3 above)
Comment (jmkulwik) [reply apps/src/applab/applab.js]: Good point. Thanks!

File: apps/src/storage/firebaseStorage.js [hunk 4]
```diff
@@ -625,21 +625,15 @@ function getRecordsData(records) {
...
-FirebaseStorage.populateTable = function(
-  jsonData,
-  overwrite,
-  onSuccess,
-  onError
-) {
+FirebaseStorage.populateTable = function(jsonData, overwrite) {
   if (!jsonData || !jsonData.length) {
-    return;
+    return Promise.resolve(true);
```
Comment (ajpal) [inline apps/src/storage/firebaseStorage.js]: Why is the `true` necessary?

Comment (jmkulwik) [reply apps/src/storage/firebaseStorage.js]: Oops. Thanks.

**Review by ajpal — APPROVED** [review]

**Pull request merged**